clickhouse-admin/src/context.rs

```diff
@@ -36,51 +67,360 @@ impl KeeperServerContext {
     pub fn log(&self) -> &Logger {
         &self.log
     }
+
+    pub fn generation_tx(&self) -> watch::Sender<Option<Generation>> {
+        self.generation_tx.clone()
+    }
 }
 
 pub struct ServerContext {
     clickhouse_cli: ClickhouseCli,
     clickward: Clickward,
-    oximeter_client: OximeterClient,
-    initialization_lock: Arc<Mutex<()>>,
+    clickhouse_address: SocketAddrV6,
     log: Logger,
+    pub generate_config_tx: Sender<GenerateConfigRequest>,
+    pub generation_tx: watch::Sender<Option<Generation>>,
+    pub db_init_tx: Sender<DbInitRequest>,
 }
 
 impl ServerContext {
-    pub fn new(clickhouse_cli: ClickhouseCli) -> Self {
-        let ip = clickhouse_cli.listen_address.ip();
-        let address = SocketAddrV6::new(*ip, CLICKHOUSE_TCP_PORT, 0, 0);
-        let oximeter_client =
-            OximeterClient::new(address.into(), &clickhouse_cli.log);
+    pub fn new(
+        log: &Logger,
+        binary_path: Utf8PathBuf,
+        listen_address: SocketAddrV6,
+    ) -> Result<Self> {
+        let clickhouse_cli =
+            ClickhouseCli::new(binary_path, listen_address, log);
+
+        let ip = listen_address.ip();
+        let clickhouse_address =
+            SocketAddrV6::new(*ip, CLICKHOUSE_TCP_PORT, 0, 0);
         let clickward = Clickward::new();
-        let log =
-            clickhouse_cli.log.new(slog::o!("component" => "ServerContext"));
-        Self {
+        let log = log.new(slog::o!("component" => "ServerContext"));
+
+        let config_path = Utf8PathBuf::from_str(CLICKHOUSE_SERVER_CONFIG_DIR)
+            .unwrap()
+            .join(CLICKHOUSE_SERVER_CONFIG_FILE);
+
+        // If there is already a configuration file with a generation number we'll
+        // use that. Otherwise, we set the generation number to None.
+        let gen = read_generation_from_file(config_path)?;
+        let (generation_tx, _rx) = watch::channel(gen);
```

Comment:
I think this is probably correct, but it isn't really making use of the `watch` channel (it's just treating it as a fancy mutex). I think what should happen here is:

* When the watch channel is created, the `tx` side is given to `long_running_generate_config_task`.
* `ServerContext` should only hold the `rx` side.
* The watch channel should not be involved in the message to generate a config at all; `long_running_generate_config_task` already has the sending side, so it can update it as needed. This means the `generate_config` endpoint no longer needs to know anything about the generation channel at all.
* The `generation` endpoint can `.borrow()` the receiving end of the channel held by `ServerContext` to read the latest value set by `long_running_generate_config_task`.